src/com/trilead/ssh2/crypto/digest/MAC.java

```diff
@@ -1,88 +1,147 @@
 
 package com.trilead.ssh2.crypto.digest;
 
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+import javax.crypto.Mac;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+
 /**
  * MAC.
  * 
  * @author Christian Plattner, plattner@trilead.com
  * @version $Id: MAC.java,v 1.1 2007/10/15 12:49:57 cplattne Exp $
  */
-public final class MAC
+public final class MAC 
 {
-	Digest mac;
-	int size;
-
-	public final static String[] getMacList()
+    private enum Hmac
 	{
-		/* Higher Priority First */
+        HMAC_MD5("hmac-md5", "HmacMD5", 16), // http://tools.ietf.org/html/rfc4253
+        HMAC_MD5_96("hmac-md5-96", "HmacMD5", 16), // http://tools.ietf.org/html/rfc4253
+        HMAC_SHA1("hmac-sha1", "HmacSHA1", 20), // http://tools.ietf.org/html/rfc4253
+        HMAC_SHA1_96("hmac-sha1-96", "HmacSHA1", 20), // http://tools.ietf.org/html/rfc4253
+        HMAC_SHA2_256("hmac-sha2-256", "HmacSHA256", 32), // http://tools.ietf.org/html/rfc6668
+        HMAC_SHA2_512("hmac-sha2-512", "HmacSHA512", 64); // http://tools.ietf.org/html/rfc6668
+
+        private String type;
+        private String algorithm;
+        private int length;
+
+        Hmac(String type, String algorithm, int length) 
+		{
+            this.type = type;
+            this.algorithm = algorithm;
+            this.length = length;
+        }
+
+        private static Hmac getHmac(String type) 
+		{
+            try 
+			{
+                return Hmac.valueOf(type.replaceAll("-", "_").toUpperCase());
+            } 
+			catch (Exception e) 
+			{
+                throw new IllegalArgumentException("Unkown algorithm " + type);
+            }
+        }
+    }
+
+    private static final String[] PRIORITIZED_MAC_LIST = { Hmac.HMAC_SHA2_256.type, Hmac.HMAC_SHA2_512.type,
+            Hmac.HMAC_SHA1_96.type, Hmac.HMAC_SHA1.type, Hmac.HMAC_MD5_96.type, Hmac.HMAC_MD5.type };
 
-		return new String[] { "hmac-sha1-96", "hmac-sha1", "hmac-md5-96", "hmac-md5" };
-	}
+    Mac mac;
```

Comment:
This is a field with the same name as one you've remove, but a different type. This may cause problems for anyone who accessed this directly.